src/auto-evo/Miche.cs

```diff
@@ -0,0 +1,210 @@
+﻿namespace AutoEvo;
+
+using System.Collections.Generic;
+using System.Linq;
+
+/// <summary>
+///   A node for the Miche Tree
+/// </summary>
+/// <remarks>
+///   <para>
+///     The Miche class forms a tree by storing a list of child instances of Miche Nodes. If a Miche has no children
+///     it is considered a leaf node and can have a species Occupant instead. This class handles insertion into the
+///     tree through scores from the selection pressure it contains. For a fuller explanation see docs/auto_evo.md
+///   </para>
+/// </remarks>
+public class Miche
+{
+    public Miche? Parent;
+    public List<Miche> Children = new();
+
+    // Occupant should always be null if this Miche has children
+    public Species? Occupant;
+    public SelectionPressure Pressure;
+
+    public Miche(SelectionPressure pressure)
+    {
+        Pressure = pressure;
+    }
+
+    public bool IsLeafNode()
+    {
+        return Children.Count == 0;
+    }
+
+    public Miche Root()
+    {
+        if (Parent == null)
+        {
+            return this;
+        }
+
+        return Parent.Root();
+    }
+
+    public IEnumerable<Miche> GetLeafNodes()
+    {
+        return GetLeafNodes(new List<Miche>());
+    }
+
+    public IEnumerable<Miche> GetLeafNodes(List<Miche> nodes)
+    {
+        if (IsLeafNode())
+        {
+            nodes.Add(this);
+            return nodes;
+        }
+
+        foreach (var child in Children)
+        {
+            nodes.AddRange(child.GetLeafNodes());
```

Comment:
Shouldn't this be `child.GetLeafNodes(nodes)`? That would avoid many memory allocations.

Reply:
That is my bad, I thought I updated that properly.